test/functional/wallet_migration.py

```diff
@@ -910,6 +908,22 @@ def test_failed_migration_cleanup(self):
             _, _, magic = struct.unpack("QII", data)
             assert_equal(magic, BTREE_MAGIC)
 
+        ####################################################
+        # Perform the same test with a loaded legacy wallet.
```

Comment:
Should we also not add the checks from lines 899 to 909 in the loaded wallet test below?
If it makes sense, then maybe we can consider calling `test_failed_migration_cleanup` twice from `run_test` with `boolean loaded_wallet` as an argument to the test.

Reply:
> Should we also not add the checks from lines 899 to 909 in the loaded wallet test below?
> If it makes sense, then maybe we can consider calling `test_failed_migration_cleanup` twice from `run_test` with `boolean loaded_wallet` as an argument to the test.

Yeah, that’s possible, but I’m not sure it’s worth the refactoring since we’re planning to remove BDB soon (this extra test will be removed too).
Also, running it from `run_test` in two different ways would require creating a different wallet for each run (because `create_legacy_wallet` can’t be called with the same name twice), which would mean decoupling the wallet name into a variable, which requires a good number of lines changes. Another option is to create a `test_failed_migration_cleanup` inner function post-setup and call it twice, e.g. 393c506533a1183225d767d18dd54c3d961e1a28.

Reply:
The approach shared in https://github.com/bitcoin/bitcoin/commit/393c506533a1183225d767d18dd54c3d961e1a28#diff-571cc226b2c9fcca7adef3916d9328f24948edf3bf0b5e66b8b45dcf75cbe502R896 looks good to me - increases assertions while getting rid of duplication as well!

Yes BDB will be removed but I'd assume the `wallet_migration` tests won't be? That's why these tests now use previous releases to create legacy wallets from what I understand: https://github.com/bitcoin/bitcoin/pull/31248

This extra test can also stick around imo, it's quite a specific case that led to the bug.

Reply:
> Yes BDB will be removed but I'd assume the `wallet_migration` tests won't be? That's why these tests now use previous releases to create legacy wallets from what I understand: #31248
> 
> This extra test can also stick around imo, it's quite a specific case that led to the bug.

Yes, `wallet_migration.py` uses a previous release binary to create the legacy wallet and the latest version binary to perform migration. Since BDB support will be removed, the latest version binary will not allow loading a legacy wallet, a prerequisite for this test, thus making it impossible to run.

Reply:
That makes sense, due to the check on line 918. Got it, thanks.